File: content/problems/3070-count-submatrices-with-top-left-element-and-sum-less-than-k/solutions/rolling-column-prefix-sums/solution.cpp
class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid,int k){
        vector<long long> columnSums(grid[0].size());
        int answer=0;
        for(auto& row:grid){
            long long prefix=0;
            for(int column=0;column<(int)row.size();++column){
                columnSums[column]+=row[column];
                prefix+=columnSums[column];
                answer+=prefix<=k;
            }
        }
        return answer;
    }
};

File: content/problems/3070-count-submatrices-with-top-left-element-and-sum-less-than-k/solutions/rolling-column-prefix-sums/Solution.kt
class Solution {
    fun countSubmatrices(grid:Array<IntArray>,k:Int):Int{
        val columnSums=LongArray(grid[0].size)
        var answer=0
        for(row in grid){
            var prefix=0L
            for(column in row.indices){
                columnSums[column]+=row[column]
                prefix+=columnSums[column]
                if(prefix<=k)answer++
            }
        }
        return answer
    }
}
